Publish iterator

tokejepsen opened the pull request:
This PR is an effort the get an iterator for publishing for iterable feedback.

This relates to https://github.com/pyblish/pyblish-nuke/pull/36.

tokejepsen committed:
Simple (but wrong) conversion to iterator.

tokejepsen commented:
This is a simple conversion, which fails a lot of tests, but its a beginning of a discussion on how to implement this.

Initially I was thinking taking the code of the ```util.publish```  into a new method ```util.publish_iterator```. ```util.publish``` would just be a simple iteration but returning the last result (the context).

Then other methods have the option to use the ```util.publish_iterator``` for interactive feedback of the publishing process.

mottosso commented:
> Initially I was thinking taking the code of the util.publish into a new method util.publish_iterator. util.publish would just be a simple iteration but returning the last result (the context).

I think that sounds like a good idea. Let's call it `publish_iter` though, to align with the keyword `iter`

tokejepsen committed:
Separate publish iterator.

tokejepsen commented:
I opted in the end for the iterator to return a tuple of the results and the context. It felt weird for a generator to return a different type of result depending on what iteration it had reached.

I still need to write a test for the publish iterator, but wanted your opinion on this first.

tokejepsen committed:
Yield percentage.

tokejepsen commented:
What do you think of the code now?

mottosso commented on pyblish/util.py:
```diff
@@ -26,13 +26,62 @@ def publish(context=None, plugins=None, targets=None):
 
     """
 
+    # Need to generate a Context object if None is passed. This is because,
```
I don't understand this? Can you give me an example?

tokejepsen replied:
If we pass in ```None``` as the context, ei. don't pass in any context to the method, the ```publish_iter``` method won't process the context object, and when we return the context object (https://github.com/pyblish/pyblish-base/pull/326/files/1306986e78f49655c131eb1787dd76d31bf4f716#diff-45fca2405e95038f14f9537f62615994R49), then it'll be ```None``` from the default value.
More specifically the two yields in ```publish_iter``` don't get executed, so the context object does not get modified. This could be resolved by always yielding the context at the end of ```publish_iter```, but then we get into the issue of yielding different results per iteration, which in my opinion makes it tricky to work with.

mottosso replied:
Aah, ok, now I understand. Yes that makes sense, but the description could use some work. In fact, I don't think you need it, it's pretty clear you need it in order to return it.

tokejepsen replied:
Fair enough, I can get rid of the comment :)

mottosso commented:
I just realised, to get a hold of the context from the iterator, at first I was thinking it would need to be yielded every time. But then it occurred to me that, the context is mutable. If you wanted a hold of it, you could make it yourself and *pass it to the iterator*.

I think this is great, because it means less to yield. But perhaps most interestingly, it means we could potentially yield *only* the result dictionary. The only obstacle to that is the percentage, which I think we could safely include for now. At least as a test. And then consider the percentage as *optional*. As in, it doesn't need to be included in results from everywhere, like the GUIs. But it can.

We could end up with this, which I think is quite elegant.

```python
for result in util.publish_iter():
  ...

# Or..
context = api.Context()
for result in util.publish_iter(context):
  ...
```

mottosso commented on pyblish/util.py:
```diff
@@ -53,8 +102,11 @@ def publish(context=None, plugins=None, targets=None):
-        plugin.process(Plugin, context, instance)
+        plugin_processed_count += 1
+        percentage = float(plugin_processed_count) / len(plugins)
+        yield (percentage, plugin.process(Plugin, context, instance), context)
```
This is risky, as `plugin.process` could throw an exception?

mottosso replied:
Ah no, sorry. I misread `plugin` for `Plugin`.

mottosso commented on pyblish/util.py:
```diff
@@ -73,16 +125,42 @@ def publish(context=None, plugins=None, targets=None):
             raise
 
-        except:  # This is unexpected, most likely a bug
+        except Exception as e:  # This is unexpected, most likely a bug
```
Can remove the `as e` and just use `throw`, unless you're going to make use of the `e`. `throw` will default to re-throwing the last thrown exception.

tokejepsen replied:
Done

mottosso commented on pyblish/util.py:
```diff
@@ -73,16 +125,42 @@ def publish(context=None, plugins=None, targets=None):
             result = plugin.process(Plugin, context, instance)
 
+            # Calculate percentage
+            instances_count = len(logic.instances_by_plugin(context, Plugin))
```
Do we need to be this exact...? Maybe.. You could also make `logic.Iterator` into a list, which would give you a fixed series of tasks to perform, and get the progress from `task_number / len(tasks)`

tokejepsen replied:
You are right, just tested it and it works better plus simplifies the code a lot. Pushed the code for a review.

mottosso commented on pyblish/util.py:
```diff
@@ -73,16 +125,42 @@ def publish(context=None, plugins=None, targets=None):
+            instances_processed_fraction = 0
+            if instances_count != 0:
+                instances_processed_fraction = (
+                    float(instances_processed_count) / instances_count
```
Great use of parentheses to combat line-length issues!

mottosso commented on pyblish/util.py:
```diff
@@ -73,16 +125,42 @@ def publish(context=None, plugins=None, targets=None):
+
+            # Increment for next iteration
+            instances_processed_count += 1
+            if plugin_processing != Plugin:
```
Yeaah, this is a little inelegant.

mottosso commented:
The beauty of an iterator in this sense is that you now control when the next task is due. This means you can include a `cancel` button and other interactive gestures in your GUI, like this.

```python
state["running"] = True

def on_clicked():
  state["running"] = False

button.clicked.connect(on_clicked)

for result in util.publish_iter():
  if not state["running"]:
    print("Cancelled")
    return

  print(result)
```

Or, even more interestingly, you can free yourself from having to use a loop. The problem with loops is that you can't stop them to do other things, like render the GUI. Running this, your button would do what it's supposed to, but it would only ever get a chance to update inbetween printing the result and the next plug-in starting to run. It'll appear very laggy to the user.

What you can do instead, is not use a loop.

```python
def on_next():
  if not state["running"]:
    print("Cancelled")
    return

  # This'll block the GUI, and run the next task
  result = next(iterator)

  # Run the next task, but take your time.
  # This is where the GUI gets a chance to be responsive
  # You can spend this time painting the GUI, for example, updating
  # The background color to signal success/failure, or populate
  # a list of things, maybe a textual or graphical summary of the result
  # dictionary from above. No end to the possibilities!
  QtCore.QTimer.singleShot(500, on_next)  

# Initialise publishing
context = api.Context()
iterator = util.publish_iter(context)

# Kick-start the process
on_next(iterator)
```

Notice there is no loop! Instead, each task is "chained" together, the last one calling the next. The beauty is that you can insert a delay to actually calling the next task! It's like a car engine. One ignition causing the next. This is impossible with a loop. It's the equivalent of calling `os.sleep()` in the middle of a loop, except you aren't actually blocking while sleeping. This is how you can get a GUI responsive, while processing things at the same time.

I took a moment to type this up, because the other day you asked about how to make a GUI responsive while busy with stuff. This is it! :)

tokejepsen committed:
Simplify percentage calculation.

mottosso commented on pyblish/util.py:
```diff
@@ -73,16 +126,22 @@ def publish(context=None, plugins=None, targets=None):
 
     # Second pass, the remainder
+    tasks = list(logic.Iterator(plugins, context, state))
     for Plugin, instance in logic.Iterator(plugins, context, state):
```
You can use the tasks you compute here, `for Plugin, instance in tasks:`

tokejepsen replied:
I actually tried that, but some tests fail:

```
======================================================================
FAIL: Decrementing order works fine
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\tje\AppData\Local\Continuum\Miniconda2\envs\tgbvfx-environment\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "C:\Users\tje\Documents\conda-git-deployment\repositories\tgbvfx-environment\pyblish-base\tests\pre13\test_logic.py", line 359, in test_decrementing_order
    assert_equals(count["#"], 111.1)
AssertionError: 1111.1 != 111.1
-------------------- >> begin captured stdout << ---------------------
I will run first, and stop things
I will stop things
I will not run

--------------------- >> end captured stdout << ----------------------

======================================================================
FAIL: The test halts an invalid publish
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\tje\AppData\Local\Continuum\Miniconda2\envs\tgbvfx-environment\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "C:\Users\tje\Documents\conda-git-deployment\repositories\tgbvfx-environment\pyblish-base\tests\pre13\test_logic.py", line 382, in test_test
    assert count["#"] == 11
AssertionError:
-------------------- >> begin captured stdout << ---------------------
I will run first, and stop things
I will not run

--------------------- >> end captured stdout << ----------------------

======================================================================
FAIL: ContextPlugin works as advertised
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\tje\AppData\Local\Continuum\Miniconda2\envs\tgbvfx-environment\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "C:\Users\tje\Documents\conda-git-deployment\repositories\tgbvfx-environment\pyblish-base\tests\test_plugin.py", line 563, in test_explicit_plugin
    assert count["#"] == 111, count
AssertionError: {'#': 1111}
-------------------- >> begin captured stdout << ---------------------
fail

--------------------- >> end captured stdout << ----------------------
-------------------- >> begin captured logging << --------------------
pyblish.Collector: INFO: Collecting from ContextPlugin
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
```

I don't quite understand it, but I think its because the iterator raises an exception https://github.com/pyblish/pyblish-base/blob/477f540981433867aa219cc46cd0fa5890ecf17c/pyblish/logic.py#L372

mottosso replied:
Ah, yes you are right.

This is another thing about iterators, while running, you can change the conditions within which it considers the next item to yield. In this case, your percentage won't be accurate. It may be "close enough", you'll have to experiment.

tokejepsen replied:
> It may be "close enough", you'll have to experiment.

I might not be understanding the processing flow correctly, but in order to get an accurate percentage I would have to predict the flow which means running the ```process``` methods twice?
I don't really mind that the percentage isn't fully accurate. You could also argue that if a validator fails, the progress bar will stop. Dunno if that is the same scenario.

tokejepsen committed:
Remove comment about context in "publish" method.

tokejepsen commented:
> I think this is great, because it means less to yield. But perhaps most interestingly, it means we could potentially yield only the result dictionary. The only obstacle to that is the percentage, which I think we could safely include for now. At least as a test. And then consider the percentage as optional. As in, it doesn't need to be included in results from everywhere, like the GUIs. But it can.

Sure, I'll take a look at this.

> I took a moment to type this up, because the other day you asked about how to make a GUI responsive while busy with stuff. This is it! :)

Thanks for taking the time on that post :) Very interesting!

tokejepsen committed:
Revert exception raising.

tokejepsen committed:
Embed percentage into yielded result.
Only yield result.

tokejepsen commented:
Are we getting to a mergable state?

mottosso commented on pyblish/util.py:
```diff
@@ -26,13 +26,59 @@ def publish(context=None, plugins=None, targets=None):
 
+    context = api.Context() if context is None else context
+
+    # Since the Context object is a singleton, we can just run though the
```
Oh, it's not a singleton. Every context is unique. There's no need for this comment either IMO.

tokejepsen replied:
Done :)

mottosso commented on pyblish/util.py:
```diff
@@ -26,13 +26,59 @@ def publish(context=None, plugins=None, targets=None):
+    Usage:
+        >> context = plugin.Context()
+        >> for result, context in util.publish_iter(context):
+               print result
```
Could you please make the print-statements Python 3 friendly? `print(result)`

tokejepsen replied:
And done :)

tokejepsen committed:
Remove incorrect singleton comment.

tokejepsen committed:
Remove redundant variable.

tokejepsen committed:
Update usage description.

mottosso commented on pyblish/util.py:
```diff
@@ -73,9 +117,15 @@ def publish(context=None, plugins=None, targets=None):
+            result["percentage"] = (
+                float(tasks_processed_count) / (len(tasks) + len(collectors))
+            )
+            yield result
```
Better yield at the very end of the loop. If you look below, you can see how unexpected exceptions are handled. At the moment, it would yield what is supposed to be a proper result before reaching that check, and yield an exception.

tokejepsen replied:
Good point.. Done :)

tokejepsen committed:
Yield result at end of for loop.

mottosso commented:
Looks good to me! If you increment the (minor) version on this, you're welcome to make a release. If possible, it would be fantastic to include a little example in the release of how and when to use it.

tokejepsen committed:
Increment minor version

tokejepsen commented:
> Looks good to me! If you increment the (minor) version on this, you're welcome to make a release. If possible, it would be fantastic to include a little example in the release of how and when to use it.

Should we cement the behaviour of ```publish_iter``` with tests, or keep it flexible while we figure out how its being used?

mottosso commented:
Oooh nice one. I would like to say I was testing you, not enforcing tests, but I hadn't even considered it. Yes, you are right, this definitely needs tests.

But, I also think you're right about us needing to figure out how we would like to use it. If you could use this to build your GUI, then that would be a good start. Then we can have a look at adding tests to keep the behaviour consistent in the future.

tokejepsen commented:
Alright, let's continue the discussion at the pyblish-nuke side and once we are happy with that PR, I'll make some tests for this to make sure the behaviour is consistent in the future.

tokejepsen committed:
Change name of "percentage" to "progress".

tokejepsen committed:
Tests for publish iteration.

tokejepsen commented:
Added the most basic tests I could think of, that needs to be in place for pyblish-nuke to function.

I'm hesitant to make more detailed tests about how the specifically the progress calculation works as we might change or update this as we figure out how its used.

What do you think?

mottosso commented:
We need tests for..

1. What happens when there are no plug-ins?
2. What happens when a bad plug-in is registered?
3. What happens when one iterator is called from two threads?

What other edge cases can you think of?

tokejepsen commented:
> What other edge cases can you think of?

Sorry, drawing a blank for this.

I'll have a look those requested tests.

mottosso commented:
The threads one applies to the vanilla `publish()` too, but I think in case of the iterator, it's more likely to be used in a GUI and one might think to use it from a thread to keep the GUI running. In Maya, this would crash. So maybe we could add to `publish()` a check that it's being called from the main thread.

tokejepsen commented:
> What happens when there are no plug-ins?

What do you expect to test with this?
When there are no plugin-ins, the iterator is just going to process nothing and any optional context passed in will be left empty.

> What happens when a bad plug-in is registered?

What do you define as a bad plug-in? Syntax issues?

tokejepsen commented:
> What happens when one iterator is called from two threads?

Very unsure about how to tackle this. Sounds like you want it to be tested in the hosts?

mottosso commented:
No problem, I'll have a closer look now.

mottosso committed:
Add "progress" to default results dictionary.

mottosso committed:
Add missing convenience iterators

Also targets cannot default to a list; lists in argument signatures are mutable, which means any change to that list will affect all calls to the function.

mottosso commented:
Putting together something with the new features, and pushed a few things I found missing. Let me know what you think.

tokejepsen commented:
All looks good to me.

Got a goal in mind with these changes?

mottosso commented:
> Got a goal in mind with these changes?

Yes, we need an example to fully validate the need for it, so I'll put together a small GUI (a little more than a splash image, but not much). As I was writing it, I found I needed `collect()` and friends, but they weren't there. I think others will come to expect it to be there too.

tokejepsen commented:
Cool, let me know when you have something to look at.

mottosso committed:
Merge branch 'master' into publish_iterator

mottosso commented:
This is a good feature and is well implemented. It's delayed because it opens so many new possibilities it'd be a shame not to document at least some of them and work up an example, but seeing as I haven't gotten around to doing that yet I figure it'd do more good out in the wild than sitting in here.

So, merging!

mottosso committed:
Merge branch 'master' into publish_iterator
